Light and Powered Light blocks

ferriarnus opened the pull request:
# Description

Light and Powered Light blocks

# Todo <!-- Remove this section if you're submitting an already-complete PR -->

- [x] Fix light updates over chunk
- [ ] Evaluate light node logic
- [x] Redstone logic
- [x] Power and wireless logic

# Checklist:

- [x] My code follows the style guidelines of this project (.editorconfig, most IDEs will use this for you)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

<!-- Thanks to: https://embeddedartistry.com/blog/2017/08/04/a-github-pull-request-template-for-your-projects/ for the building blocks of this template -->

ferriarnus committed:
Light works somewhat

light updates need to be fixed for chunk borders
Need the actual "power" part
Need inverted and redstone logic

ferriarnus committed:
Redstone logic

ferriarnus commented on build.gradle:
```diff
@@ -39,7 +39,7 @@ subprojects {
     def enableMixin = false
     if (project.hasProperty("enable_mixin")) {
-        enableMixin = project.enable_mixin
+        //enableMixin = project.enable_mixin
```
need to fix this, eclipse just can't handle mixins

ferriarnus commented:
should this be merged before #46, or should it be moved to base already and merged after it.

Rover656 commented:
> should this be merged before #46, or should it be moved to base already and merged after it.

I'd say merge this first, #46 still has a bit of work to do that isn't finished yet, so I'll tidy that branch up when I get back to it :)

ferriarnus committed:
Merge branch 'dev/1.18.x' into lights

ferriarnus committed:
Let there be light

Finally lights work (tm)
Some cleanup will be needed, especially the network stuff, since this will be joined in base.

ferriarnus committed:
Update PoweredLightBlockEntity.java

oops

Rover656 commented on enderio-decoration/src/main/java/com/enderio/decoration/common/block/light/Light.java:
```diff
@@ -0,0 +1,97 @@
+	protected static final VoxelShape EAST_AABB = Block.box(0.0D, 6.0D, 5.0D, 2.0D, 10.0D, 11.0D);
+	private boolean inverted;
+	
+	public Light(Properties p_53182_, boolean inverted) {
```
minor: Rename the mojmapped param name

Rover656 commented on enderio-decoration/src/main/java/com/enderio/decoration/common/block/light/Light.java:
```diff
@@ -0,0 +1,97 @@
+import net.minecraft.world.phys.shapes.CollisionContext;
+import net.minecraft.world.phys.shapes.VoxelShape;
+
+public class Light extends FaceAttachedHorizontalDirectionalBlock{
```
A minor documentation pass could be good.

Rover656 commented on enderio-decoration/src/main/java/com/enderio/decoration/common/block/light/LightNode.java:
```diff
@@ -0,0 +1,58 @@
+
+public class LightNode extends Block implements EntityBlock{
+
+	public LightNode(Properties p_49224_) {
```
Param name :)

Rover656 commented on enderio-decoration/src/main/java/com/enderio/decoration/common/block/light/PoweredLight.java:
```diff
@@ -0,0 +1,68 @@
+public class PoweredLight extends Light implements EntityBlock{
+	public boolean wireless;
+
+	public PoweredLight(Properties p_49224_, boolean inverted, boolean wireless) {
```
Param name :)

Rover656 commented on enderio-decoration/src/main/java/com/enderio/decoration/common/blockentity/LightNodeBlockEntity.java:
```diff
@@ -0,0 +1,76 @@
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraftforge.network.PacketDistributor;
+
+public class LightNodeBlockEntity extends BlockEntity{
```
Space after BlockEntity :)

Rover656 commented on enderio-decoration/src/main/java/com/enderio/decoration/common/blockentity/PoweredLightBlockEntity.java:
```diff
@@ -0,0 +1,208 @@
+	/**
+	 * Add Light nodes to the axis of the center block. If the position is already occupied, it is added to the blocked list. 
+	 * If the position is not blocked by a block in the blocked list and is within the spread zone, it is added to the array of starting positions for the next iteration.
+	 * @param level
```
I think dropping unused things like these param's are better for readability

Rover656 commented on enderio-decoration/src/main/java/com/enderio/decoration/common/blockentity/PoweredLightBlockEntity.java:
```diff
@@ -0,0 +1,208 @@
+		}
+	}
+	
+	private static void consumePowerWireless(Level level, BlockPos pos, BlockState state, PoweredLightBlockEntity e) {
```
If I'm reading this right, this doesn't consume any power?

ferriarnus replied:
Yes, since we do not have the wireless version. I should have put a comment there. This is just the behavior needed to set the light on.

Rover656 replied:
Alrighty, sounds good :) Might be worth writing some power consumption utilities in base/core when we do implement wireless so we can reuse the capability elsewhere later.

Rover656 replied:
Still needs a todo here :)

Rover656 commented on enderio-decoration/src/main/java/com/enderio/decoration/common/network/EnderDecorNetwork.java:
```diff
@@ -0,0 +1,23 @@
+        INSTANCE.registerMessage(0, ServerToClientLightUpdate.class, ServerToClientLightUpdate::write, ServerToClientLightUpdate::new, ServerToClientLightUpdate::handle);
+    }
+
+    
```
Double newline

Rover656 commented on enderio-decoration/src/main/java/com/enderio/decoration/common/network/ServerToClientLightUpdate.java:
```diff
@@ -0,0 +1,37 @@
+    }
+    
+    static void handle(final ServerToClientLightUpdate message, Supplier<Context> ctx) {
+        ctx.get().enqueueWork(() ->
```
Could this all be on one line?

ferriarnus replied:
I will change it, however, this packet stuff will need to be merged with the one in base, so I might rewrite it a little to match that. I just needed a quick and dirty channel for this.

Rover656 replied:
Thats no problem at all, I'll handle the rest once this is merged :)

Rover656 commented on enderio-decoration/src/main/java/com/enderio/decoration/common/network/ServerToClientLightUpdateHandler.java:
```diff
@@ -0,0 +1,14 @@
+public class ServerToClientLightUpdateHandler {
+
+    public static void handlePacket(ServerToClientLightUpdate message, Supplier<Context> ctx) {
+        Minecraft.getInstance().level.setBlock(message.pos, message.state, 3);
```
Could this be done without the need of sending the state over the network and instead querying for the state already present? Might save on traffic.

ferriarnus replied:
We will need a packet either way, since the state is set from the server. I currently only send this packet to the server, since the flag is 1 on the server. But that needs to be double checked.

Rover656 replied:
Yeah, I mean rather than sending the state as part of the message could we not just save the position and find the state based on said position, it'd make the packet smaller.

ferriarnus replied:
But that means the state is already known on the client, which isn't the case. The client does not have the block (if i read the flags right).

Rover656 commented on enderio-decoration/src/main/java/com/enderio/decoration/datagen/model/block/DecorBlockState.java:
```diff
@@ -19,4 +26,21 @@ public static void paintedBlock(DataGenContext<Block, ? extends Block> ctx, Regi
     }
+    
+    public static void lightBlock(DataGenContext<Block, ? extends Block> ctx, RegistrateBlockstateProvider prov) {
+    	prov.getVariantBuilder(ctx.get()).forAllStates(state -> {
```
Look at your formatting in this method :)

ferriarnus replied:
I'm not to sure what you mean?

Rover656 reviewed (changes requested):
Just a breif scan over, not able to run this branch right now but just some code clarity comments and a couple of questions :)

ferriarnus committed:
first round of fixes

ferriarnus committed:
first doc pass

ferriarnus committed:
Update DecorBlockState.java

ferriarnus committed:
Update PoweredLightBlockEntity.java

ferriarnus committed:
Merge branch 'dev/1.18.x' into lights

Rover656 reviewed (approved):
Looking good, good work! Thanks!
